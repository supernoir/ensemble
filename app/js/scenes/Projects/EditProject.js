import React from 'react';
import intl from 'react-intl-universal';
import { Link } from 'react-router-dom';
import { Container, Segment, Form, Button, Divider, Header, Breadcrumb } from 'semantic-ui-react';
import Loader from '../../basics/Loader';
import { tagTypes } from '../../constants/tagTypes';
import { projectTypes } from '../../constants/projectTypes';
import { projectStatus } from '../../constants/projectStatus';

/**
 * Class EditProject
 */
export default class EditProject extends React.Component {
	constructor() {
		super();
		this.state = {
			title        : '',
			status       : '',
			genre        : '',
			series       : '',
			cast         : '',
			desc         : '',
			tags         : [],
			projectTypes : [],
			projectStatus: [],
			project      : {}
		};
	}

	handleInput = (source, evt) => {
		switch (source) {
			case 'type':
				this.setState({ type: evt.currentTarget.value });
				break;
			case 'status':
				this.setState({ status: evt.currentTarget.value });
				break;
			case 'title':
				this.setState({ title: evt.currentTarget.value });
				break;
			case 'series':
				this.setState({ series: evt.currentTarget.value });
				break;
			case 'genre':
				this.setState({ genre: evt.currentTarget.value });
				break;
			case 'desc':
				this.setState({ desc: evt.currentTarget.value });
				break;
			case 'cast':
				this.setState({ cast: evt.currentTarget.value });
				break;
			case 'tags':
				this.setState({
					tags: evt.currentTarget.value.split(',').map(item => item.trim())
				});
				break;
			default:
				break;
		}
	}

	postEditedProject(evt) {
		evt.preventDefault();

		this.props.editProject(this.props.match.params.id, {
			type  : this.state.type || this.props.project.type,
			status: this.state.status || this.props.project.status || projectStatus.DRAFT,
			title : this.state.title || this.props.project.title,
			genre : this.state.genre || this.props.project.genre,
			series: this.state.series || this.props.project.series,
			desc  : this.state.desc || this.props.project.desc,
			cast  : this.state.cast || this.props.project.cast,
			tags  : this.state.tags || this.props.project.tags
		});

		if (this.state.tags !== void 0) {
			this.state.tags.forEach(tag => {
				this.props.addTag({
					type: tagTypes.PROJECT,
					name: tag,
					ref : this.state.title
				});
			});
		}

		this.props.addTag({
			type: tagTypes.PROJECT_TYPE,
			name: this.state.type,
			ref : this.state.title
		});

		this.props.addEvent({
			user  : 'testAdmin',
			action: 'edit_project',
			ref   : this.state.title
		});

		this.props.history.push('/projects');
	}

	componentWillMount = () => {
		this.props.getProjectById(this.props.match.params.id);

		let types = [];
		let statuses = [];
		for (let type in projectTypes) {
			types.push(projectTypes[type]);
		}
		for (let status in projectStatus) {
			statuses.push(projectStatus[status]);
		}
		this.setState({
			projectTypes : types,
			projectStatus: statuses
		});
	}

	render() {
		return this.props.loading
			? <Loader loading={this.props.loading} />
			: <Container>
				<Breadcrumb>
					<Breadcrumb.Section link>
						<Link to="/">{intl.get('component.dashboard')}</Link>
					</Breadcrumb.Section>
					<Breadcrumb.Divider />
					<Breadcrumb.Section link>
						<Link to="/projects">{intl.get('entity.projects')}</Link>
					</Breadcrumb.Section>
					<Breadcrumb.Divider />
					<Breadcrumb.Section active>
						{intl.get('project.action-edit')}
					</Breadcrumb.Section>
				</Breadcrumb>

				<Segment>
					<Header as="h2">
						{intl.get('project.action-edit')}
					</Header>

					<Divider />

					<Form>
						<Form.Field>
							<label htmlFor="type" className="col-sm-2 control-label">{intl.get('project.label-type')}</label>
							<select name="type" className={'ui search selection dropdown'} onChange={evt => this.handleInput('type', evt)}>
								{this.state.projectTypes !== void 0 && this.state.projectTypes.length > 0
									? this.state.projectTypes.map((type, index) => {
										return (
											<option key={`${type}-${index}`} value={type}>
												{intl.get(`project.select-type-${type}`)}
											</option>
										);
									})
									: null}
							</select>
						</Form.Field>
						<Form.Field>
							<label htmlFor="status" className="col-sm-2 control-label">{intl.get('project.label-status')}</label>
							<select name="status" className={'ui search selection dropdown'} onChange={evt => this.handleInput('status', evt)}>
								{this.state.projectStatus !== void 0 && this.state.projectStatus.length > 0
									? this.state.projectStatus.map((status, index) => {
										return (
											<option key={`${status}-${index}`} value={status}>
												{intl.get(`project.status-${status}`)}
											</option>
										);
									})
									: null}
							</select>
						</Form.Field>
						<Form.Field>
							<label for="project" className="col-sm-2 control-label">{intl.get('project.label-title')}</label>
							<input
								onChange={evt => this.handleInput('title', evt)}
								type="text"
								className="form-control"
								id="title"
								placeholder="The Final Problem"
								defaultValue={this.props.project.title}
								required
							/>
						</Form.Field>
						<Form.Field>
							<label for="project" className="col-sm-2 control-label">{intl.get('project.label-genre')}</label>
							<input
								onChange={evt => this.handleInput('genre', evt)}
								type="text"
								className="form-control"
								id="genre"
								placeholder="Crime, Suspense"
								defaultValue={this.props.project.genre}
							/>
						</Form.Field>
						<Form.Field>
							<label for="project" className="col-sm-2 control-label">{intl.get('project.label-series')}</label>
							<input
								onChange={evt => this.handleInput('series', evt)}
								type="text"
								className="form-control"
								id="series"
								placeholder="The Memoirs of Sherlock Holmes"
								defaultValue={this.props.project.series}
							/>
						</Form.Field>
						<Form.Field>
							<label for="project" className="col-sm-2 control-label">{intl.get('project.label-cast')}</label>
							<input
								onChange={evt => this.handleInput('cast', evt)}
								type="text"
								className="form-control"
								id="cast"
								placeholder="Sherlock Holmes, James Watson"
								defaultValue={this.props.project.cast}
							/>
						</Form.Field>
						<Form.Field>
							<label for="project" className="col-sm-2 control-label">{intl.get('project.label-description')}</label>
							<input
								onChange={evt => this.handleInput('desc', evt)}
								type="text"
								className="form-control"
								id="desc"
								placeholder="The Adventures of Sherlock Holmes"
								defaultValue={this.props.project.desc}
								required
							/>
						</Form.Field>
						<Form.Field>
							<label for="project" className="col-sm-2 control-label">{intl.get('project.label-tags')}</label>
							<input
								onChange={evt => this.handleInput('tags', evt)}
								type="text"
								className="form-control"
								id="tags"
								placeholder="crime, noir"
								defaultValue={this.props.project.tags !== void 0 ? this.props.project.tags.join(', ') : ''}
							/>
						</Form.Field>

						<Button
							onClick={evt => this.postEditedProject(evt)}
							type="submit"
							className="btn btn-default">{intl.get('project.action-edit')}</Button>
					</Form>

				</Segment>
			</Container>;
	}
}
