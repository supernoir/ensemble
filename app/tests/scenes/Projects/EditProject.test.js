import React from 'react';
import { shallow } from 'enzyme';

// Import Mocks
import mockProject from '../../mocks/mockProject';

// Import Components
import EditProject from '../../../js/scenes/Projects/EditProject';

test('EditProject -- Snapshot Test', () => {
	const wrapper = shallow(
		<EditProject
			loading={false}
			editProject={(id, data) => {
				jest.fn(id, data);
			}}
			getProjectById={id => jest.fn(id)}
			project={mockProject}
			addEvent={data => {
				jest.fn(data);
			}}
			addTag={data => {
				jest.fn(data);
			}}
			match={{ params: 1 }}
		/>
	);
	expect(wrapper).toMatchSnapshot();
});
