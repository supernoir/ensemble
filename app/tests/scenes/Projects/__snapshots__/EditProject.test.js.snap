// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProject -- Snapshot Test 1`] = `
<Container>
  <Breadcrumb>
    <BreadcrumbSection>
      <Link
        replace={false}
        to="/"
      />
    </BreadcrumbSection>
    <BreadcrumbDivider />
    <BreadcrumbSection>
      <Link
        replace={false}
        to="/projects"
      />
    </BreadcrumbSection>
    <BreadcrumbDivider />
    <BreadcrumbSection
      active={true}
    />
  </Breadcrumb>
  <Segment>
    <Header
      as="h2"
    />
    <Divider />
    <Form
      as="form"
    >
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="type"
        />
        <select
          className="ui search selection dropdown"
          name="type"
          onChange={[Function]}
        >
          <option
            key="shortstory-0"
            value="shortstory"
          />
          <option
            key="script-1"
            value="script"
          />
          <option
            key="novel-2"
            value="novel"
          />
          <option
            key="essay-3"
            value="essay"
          />
          <option
            key="poem-4"
            value="poem"
          />
        </select>
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="status"
        />
        <select
          className="ui search selection dropdown"
          name="status"
          onChange={[Function]}
        >
          <option
            key="draft-0"
            value="draft"
          />
          <option
            key="inprogress-1"
            value="inprogress"
          />
          <option
            key="inreview-2"
            value="inreview"
          />
          <option
            key="editing-3"
            value="editing"
          />
          <option
            key="published-4"
            value="published"
          />
        </select>
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="project"
        />
        <input
          className="form-control"
          defaultValue="The Passersby"
          id="title"
          onChange={[Function]}
          placeholder="The Final Problem"
          required={true}
          type="text"
        />
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="project"
        />
        <input
          className="form-control"
          defaultValue="Drama"
          id="genres"
          onChange={[Function]}
          placeholder="Crime, Suspense"
          type="text"
        />
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="project"
        />
        <input
          className="form-control"
          defaultValue="The Passersby"
          id="series"
          onChange={[Function]}
          placeholder="The Memoirs of Sherlock Holmes"
          type="text"
        />
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="project"
        />
        <input
          className="form-control"
          defaultValue="David, Sean"
          id="cast"
          onChange={[Function]}
          placeholder="Sherlock Holmes, James Watson"
          type="text"
        />
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="project"
        />
        <input
          className="form-control"
          defaultValue="Dublin story"
          id="desc"
          onChange={[Function]}
          placeholder="The Adventures of Sherlock Holmes"
          required={true}
          type="text"
        />
      </FormField>
      <FormField>
        <label
          className="col-sm-2 control-label"
          htmlFor="project"
        />
        <input
          className="form-control"
          defaultValue="example, sample, test"
          id="tags"
          onChange={[Function]}
          placeholder="crime, noir"
          type="text"
        />
      </FormField>
      <Button
        as="button"
        className="btn btn-default"
        onClick={[Function]}
        type="submit"
      />
    </Form>
  </Segment>
</Container>
`;
